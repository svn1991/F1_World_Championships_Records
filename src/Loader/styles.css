#loading-wrapper {
  width: 100%;
  height: 125px;
  background: white;
  position: relative;
}

#loading-text {
  display: block;
  position: absolute;
  top: 50%;
  left: 50%;
  color: #999;
  width: 100px;
  height: 30px;
  margin: -18px 0 0 -48px;
  text-align: center;
  font-family: "PT Sans Narrow", sans-serif;
  font-size: 10px;
}

#loading-content {
  display: block;
  position: relative;
  width: 100px;
  height: 100px;
  margin: 15px auto;
}

#loading-content:after {
  content: "";
  position: absolute;
  border: 3px solid #0f0;
  left: 15px;
  right: 15px;
  top: 15px;
  bottom: 15px;
}

#loading-content:before {
  content: "";
  position: absolute;
  border: 3px solid #00f;
  left: 5px;
  right: 5px;
  top: 5px;
  bottom: 5px;
}

#loading-content {
  border: 3px solid transparent;
  border-top-color: #4d658d;
  border-bottom-color: #4d658d;
  border-radius: 50%;
  -webkit-animation: loader 2s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
  animation: loader 2s linear infinite;
}

#loading-content:before {
  border: 3px solid transparent;
  border-top-color: #d4cc6a;
  border-bottom-color: #d4cc6a;
  border-radius: 50%;
  -webkit-animation: loader 3s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
  animation: loader 3s linear infinite;
}

#loading-content:after {
  border: 3px solid transparent;
  border-top-color: #84417c;
  border-bottom-color: #84417c;
  border-radius: 50%;
  -webkit-animation: loader 1.5s linear infinite;
  animation: loader 1.5s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
}

@-webkit-keyframes loaders {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes loader {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
