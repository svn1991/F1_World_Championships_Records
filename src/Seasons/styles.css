.seasons-wrapper {
  max-width: 920px;
  margin: 50px auto;
}
.seasons-wrapper .seasons-title {
  margin-bottom: 30px;
}

.seasons-wrapper #loading-wrapper {
  margin: 110px 0;
}
