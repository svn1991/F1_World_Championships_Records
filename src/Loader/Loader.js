import React from 'react';
import './styles.css';

const Loader = () => (
  <div id="loading-wrapper">
    <div id="loading-text">LOADING</div>
    <div id="loading-content" />
  </div>
);

export default Loader;
