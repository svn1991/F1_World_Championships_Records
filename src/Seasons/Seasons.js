import React, { useState, useEffect } from "react";
import { getSeasonsInfoFrom, getWinnerForYear } from "../CacheData/CacheData";
import BootstrapTable from "react-bootstrap-table-next";
import PropTypes from "prop-types";

import Races from "../Races/Races";
import Loader from "../Loader/Loader";

import "./styles.css";

function Seasons(props) {
  const { startYear, endYear } = props;
  const [seasons, setSeasons] = useState({});

  /**
   * Get season and winner name information from API
   */
  useEffect(() => {
    let isSubscribed = true;
    getSeasonsInfoFrom(startYear, endYear).then(seasons => {
      const winnerPromises = seasons.map(valueObject =>
        getWinnerForYear(valueObject.season)
      );
      Promise.all(winnerPromises).then(winners => {
        const seasonObject = {};
        winners.forEach((winner, index) => {
          const season = seasons[index];
          seasonObject[season.season] = {
            winner,
            year: season.season,
            url: season.url
          };
        });
        if (isSubscribed) {
          setSeasons(seasonObject);
        }
      });
    });
    return () => (isSubscribed = false);
  }, []);

  /**
   * Get season and winner information organised into
   * their rows and column titles
   */
  function getSeasonsProducts() {
    const years = Object.keys(seasons);
    return years.map(year => {
      return {
        id: year,
        winnerId: seasons[year].winner.driverId,
        season: <a href={seasons[year].url}>{year}</a>,
        winner: (
          <a href={seasons[year].winner.url}>
            {`${seasons[year].winner.givenName} ${
              seasons[year].winner.familyName
            }`}
          </a>
        )
      };
    });
  }

  /**
   * Column headers
   */
  const columns = [
    {
      dataField: "season",
      text: "Season"
    },
    {
      dataField: "winner",
      text: "Winner"
    }
  ];

  const expandRow = {
    onlyOneExpanding: true,
    showExpandColumn: true,
    renderer: row => <Races year={row.id} seasonWinner={row.winnerId} />,
    expandHeaderColumnRenderer: ({ isAnyExpands }) => {
      if (isAnyExpands) {
        return <b>-</b>;
      }
      return <b>+</b>;
    },
    expandColumnRenderer: ({ expanded }) => {
      if (expanded) {
        return <b>-</b>;
      }
      return <b>...</b>;
    }
  };

  return (
    <div className="seasons-wrapper">
      <h1 className="seasons-title">
        {`F1 World Championships: ${startYear}-${endYear}`}
      </h1>
      {Object.keys(seasons).length === 0 ? (
        <Loader />
      ) : (
        <BootstrapTable
          keyField="id"
          data={getSeasonsProducts()}
          columns={columns}
          expandRow={expandRow}
          bordered={false}
          condensed
        />
      )}
    </div>
  );
}

Seasons.propTypes = {
  startYear: PropTypes.number.isRequired,
  endYear: PropTypes.number.isRequired
};

export default Seasons;
