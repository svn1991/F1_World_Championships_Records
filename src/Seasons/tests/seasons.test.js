import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Seasons from '../Seasons';
import Loader from '../../Loader/Loader';
import * as SeasonsData from './seasons-2001-2015.json';
import * as WinnersData from './season-winners-2001-2015.json';

const seasonsInfo = SeasonsData.MRData.SeasonTable.Seasons;
const winnersInfo = WinnersData.MRData.StandingsTable.StandingsLists;

const window = global;
const storage = window.localStorage;

describe('<Seasons />', () => {
  beforeEach(() => {
    winnersInfo.forEach(winnerInfo => {
      storage.setItem(`https://ergast.com/api/f1/${winnerInfo.season}/driverStandings/1.json`, JSON.stringify({
        ...winnerInfo.DriverStandings[0].Driver,
        winningYearRecord: winnerInfo.season,
      }));
    });
  });
  afterEach(() => {
    storage.clear();
    cleanup();
  });

  it.skip('Expect to log errors in console', done => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Seasons />);
    requestAnimationFrame(() => {
      expect(spy).toHaveBeenCalled();
      done();
    });
  });

  // FIXME: correct this test, act render is having an issue
  it.skip('Expect to not log errors in console', done => {
    storage.setItem('https://ergast.com/api/f1/seasons.json?limit=15&offset=51', JSON.stringify([]));
    const spy = jest.spyOn(global.console, 'error');
    render(<Seasons startYear={2001} endYear={2015} />);
    requestAnimationFrame(() => {
      expect(spy).not.toHaveBeenCalled();
      done();
    });
  });

  it('should render and match the snapshot', done => {
    storage.setItem('https://ergast.com/api/f1/seasons.json?limit=15&offset=51', JSON.stringify([]));
    const { container: { firstChild } } = render(<Seasons startYear={2001} endYear={2015} />);
    requestAnimationFrame(() => {
      expect(firstChild).toMatchSnapshot();
      done();
    });
  });

  it('should render and match the snapshot with seasons data', done => {
    storage.setItem('https://ergast.com/api/f1/seasons.json?limit=15&offset=51', JSON.stringify(seasonsInfo));
    const { container } = render(<Seasons startYear={2001} endYear={2015} />);

    requestAnimationFrame(() => {
      expect(container.firstChild).toMatchSnapshot();
      done();
    });
  });

  it('should mount loader component when no season information is available', done => {
    storage.setItem('https://ergast.com/api/f1/seasons.json?limit=15&offset=51', JSON.stringify([]));
    const { container } = render(<Seasons startYear={2001} endYear={2015} />);
    const loader = render(<Loader />);

    requestAnimationFrame(() => {
      const seasonLoading = container.querySelector('.seasons-wrapper #loading-wrapper');
      const loadContainer = loader.container.querySelector('#loading-wrapper');
      expect(seasonLoading).toEqual(loadContainer);
      done();
    });
  });

  it('random check season row to have season year, url, winner, winner url displayed, and click loads races info', done => {
    storage.setItem('https://ergast.com/api/f1/seasons.json?limit=15&offset=51', JSON.stringify(seasonsInfo));
    const { container } = render(<Seasons startYear={2001} endYear={2015} />);

    requestAnimationFrame(() => {
      expect(container.querySelectorAll('.races-wrapper').length).toEqual(0);
      const tableBodyRows = container.querySelectorAll('table tbody tr');

      const season = tableBodyRows[3];
      expect(season).toEqual(expect.anything());
      const columns = season.children;
      expect(columns[1].children[0].innerHTML).toEqual('2004');
      expect(columns[1].children[0].href).toEqual('https://en.wikipedia.org/wiki/2004_Formula_One_season');

      expect(columns[2].children[0].innerHTML).toEqual('Michael Schumacher');
      expect(columns[2].children[0].href).toEqual('http://en.wikipedia.org/wiki/Michael_Schumacher');

      fireEvent.click(season);
      expect(container.querySelectorAll('.races-wrapper').length).toEqual(1);
      done();
    });
  });
});
