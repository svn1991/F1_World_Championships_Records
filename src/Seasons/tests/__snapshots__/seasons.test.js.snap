// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Seasons /> should render and match the snapshot 1`] = `
<div
  class="seasons-wrapper"
>
  <h1
    class="seasons-title"
  >
    F1 World Championships: 2001-2015
  </h1>
  <div
    id="loading-wrapper"
  >
    <div
      id="loading-text"
    >
      LOADING
    </div>
    <div
      id="loading-content"
    />
  </div>
</div>
`;

exports[`<Seasons /> should render and match the snapshot with seasons data 1`] = `
<div
  class="seasons-wrapper"
>
  <h1
    class="seasons-title"
  >
    F1 World Championships: 2001-2015
  </h1>
  <div
    class="react-bootstrap-table"
  >
    <table
      class="table table-condensed"
    >
      <thead>
        <tr>
          <th
            data-row-selection="true"
          >
            <b>
              +
            </b>
          </th>
          <th
            tabindex="0"
          >
            Season
          </th>
          <th
            tabindex="0"
          >
            Winner
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2001_Formula_One_season"
            >
              2001
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Michael_Schumacher"
            >
              Michael Schumacher
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2002_Formula_One_season"
            >
              2002
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Michael_Schumacher"
            >
              Michael Schumacher
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2003_Formula_One_season"
            >
              2003
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Michael_Schumacher"
            >
              Michael Schumacher
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2004_Formula_One_season"
            >
              2004
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Michael_Schumacher"
            >
              Michael Schumacher
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2005_Formula_One_season"
            >
              2005
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            >
              Fernando Alonso
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2006_Formula_One_season"
            >
              2006
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Fernando_Alonso"
            >
              Fernando Alonso
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2007_Formula_One_season"
            >
              2007
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
            >
              Kimi Räikkönen
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2008_Formula_One_season"
            >
              2008
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Lewis_Hamilton"
            >
              Lewis Hamilton
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2009_Formula_One_season"
            >
              2009
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Jenson_Button"
            >
              Jenson Button
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2010_Formula_One_season"
            >
              2010
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
            >
              Sebastian Vettel
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2011_Formula_One_season"
            >
              2011
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
            >
              Sebastian Vettel
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2012_Formula_One_season"
            >
              2012
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
            >
              Sebastian Vettel
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2013_Formula_One_season"
            >
              2013
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
            >
              Sebastian Vettel
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2014_Formula_One_season"
            >
              2014
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Lewis_Hamilton"
            >
              Lewis Hamilton
            </a>
          </td>
        </tr>
        <tr
          class=""
        >
          <td>
            <b>
              ...
            </b>
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/2015_Formula_One_season"
            >
              2015
            </a>
          </td>
          <td>
            <a
              href="http://en.wikipedia.org/wiki/Lewis_Hamilton"
            >
              Lewis Hamilton
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
